File: src/generation/delete.zig
const std = @import("std");
const opts = @import("options");
const fmt = std.fmt;
const fs = std.fs;
const io = std.io;
const mem = std.mem;
const linux = std.os.linux;
const posix = std.posix;
const sort = std.sort;
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const ArgIterator = std.process.ArgIterator;

const argparse = @import("../argparse.zig");
const ArgParseError = argparse.ArgParseError;
const argError = argparse.argError;
const argIs = argparse.argIs;
const getNextArgs = argparse.getNextArgs;

const Constants = @import("../constants.zig");

const log = @import("../log.zig");

const zeit = @import("zeit");

const utils = @import("../utils.zig");
const GenerationMetadata = utils.generation.GenerationMetadata;
const TimeSpan = utils.time.TimeSpan;
const runCmd = utils.runCmd;

pub const GenerationDeleteCommand = struct {
    all: bool = false,
    from: ?usize = null,
    keep: ArrayList(usize),
    min: ?usize = null,
    older_than: ?TimeSpan = null,
    to: ?usize = null,
    yes: bool = false,
    remove: ArrayList(usize), // Positional args, not using an option

    const Self = @This();

    const usage =
        \\Delete NixOS generations from this system.
        \\
        \\Usage:
        \\    nixos generation delete [GEN...] [options]
        \\
        \\Arguments:
        \\    [GEN...]    Generation numbers to delete
        \\
        \\Options:
        \\    -a, --all                    Delete all generations except the current one
        \\    -f, --from <GEN>             Delete all generations after <GEN>, inclusive
        \\    -h, --help                   Show this help menu
        \\    -k, --keep <GEN>             Always keep this generation; can be used multiple times
        \\    -m, --min <NUM>              Keep a minimum of <NUM> generations
        \\    -o, --older-than <PERIOD>    Delete all generations older than <PERIOD>
        \\    -t, --to <GEN>               Delete this generation
        \\    -y, --yes                    Automatically confirm deletion of generations
        \\
        \\Values:
        \\    <GEN>       Generation number
        \\    <PERIOD>    systemd.time span (i.e. "30d 2h 1m")
        \\
        \\These options and arguments can be combined ad-hoc.
        \\
    ;
    // TODO: add manpage examples

    fn parseGenNumber(candidate: []const u8) !usize {
        return fmt.parseInt(usize, candidate, 10) catch {
            argError("'{s}' is not a valid generation number", .{candidate});
            return ArgParseError.InvalidArgument;
        };
    }

    pub fn init(allocator: Allocator) Self {
        return GenerationDeleteCommand{
            .keep = ArrayList(usize).init(allocator),
            .remove = ArrayList(usize).init(allocator),
        };
    }

    pub fn parseArgs(argv: *ArgIterator, parsed: *GenerationDeleteCommand) !?[]const u8 {
        var next_arg = argv.next();
        while (next_arg) |arg| {
            if (argIs(arg, "--all", "-a")) {
                parsed.all = true;
            } else if (argIs(arg, "--from", "-f")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                parsed.from = try parseGenNumber(next);
            } else if (argIs(arg, "--help", "-h")) {
                log.print("{s}", .{usage});
                return ArgParseError.HelpInvoked;
            } else if (argIs(arg, "--keep", "-k")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                try parsed.keep.append(try parseGenNumber(next));
            } else if (argIs(arg, "--min", "-m")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                parsed.min = fmt.parseInt(usize, next, 10) catch {
                    argError("'{s}' is not a number", .{next});
                    return ArgParseError.InvalidArgument;
                };
                if (parsed.min == 0) {
                    argError("--min must be at least 1", .{});
                    return ArgParseError.InvalidArgument;
                }
            } else if (argIs(arg, "--older-than", "-o")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                parsed.older_than = TimeSpan.fromSlice(next) catch |err| {
                    argError("'{s}' is not formatted correctly: {s}", .{ next, @errorName(err) });
                    return ArgParseError.InvalidArgument;
                };
            } else if (argIs(arg, "--to", "-t")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                parsed.to = try parseGenNumber(next);
            } else if (argIs(arg, "--yes", "-y")) {
                parsed.yes = true;
            } else {
                if (argparse.isFlag(arg)) {
                    return arg;
                }

                try parsed.remove.append(try parseGenNumber(arg));
            }

            next_arg = argv.next();
        }

        for (parsed.remove.items) |remove| {
            for (parsed.keep.items) |keep| {
                if (remove == keep) {
                    argError("generation {d} cannot be both removed and kept", .{remove});
                    return ArgParseError.ConflictingOptions;
                }
            }
        }

        if (parsed.all) {
            if (parsed.from != null) {
                log.warn("--all was specified, ignoring --from", .{});
            }
            if (parsed.older_than != null) {
                log.warn("--all was specified, ignoring --older-than", .{});
            }
            if (parsed.to != null) {
                log.warn("--all was specified, ignoring --to", .{});
            }
            if (parsed.remove.items.len > 0) {
                log.warn("--all was specified, ignoring positional arguments", .{});
            }
        }

        if (!parsed.all and
            parsed.from == null and
            parsed.keep.items.len == 0 and
            parsed.min == null and
            parsed.older_than == null and
            parsed.to == null and
            parsed.remove.items.len == 0)
        {
            argError("no generations or deletion parameters were given", .{});
            return ArgParseError.MissingRequiredArgument;
        }

        return null;
    }

    pub fn deinit(self: Self) void {
        self.keep.deinit();
        self.remove.deinit();
    }
};

pub const GenerationDeleteError = error{
    CurrentGenerationRequested,
    ResourceAccessFailed,
    InvalidParameter,
    PermissionDenied,
    CommandFailed,
} || Allocator.Error;

pub fn printDeleteSummary(allocator: Allocator, generations: []GenerationMetadata) !void {
    log.print("The following generations will be deleted:\n\n", .{});

    const headers: []const []const u8 = &.{ "#", "Description", "Creation Date" };

    var max_row_len = comptime blk: {
        var tmp: []const usize = &[_]usize{};
        for (headers) |header| {
            tmp = tmp ++ [_]usize{header.len};
        }
        var new: [tmp.len]usize = undefined;
        std.mem.copyForwards(usize, &new, tmp);
        break :blk new;
    };

    var num_list = try ArrayList([]const u8).initCapacity(allocator, generations.len);
    defer {
        for (num_list.items) |num| allocator.free(num);
        num_list.deinit();
    }

    var date_list = try ArrayList([]const u8).initCapacity(allocator, generations.len);
    defer {
        for (date_list.items) |date| allocator.free(date);
        date_list.deinit();
    }

    for (generations) |gen| {
        const t = gen.date.?;

        const num = try fmt.allocPrint(allocator, "{d}", .{gen.generation.?});
        num_list.appendAssumeCapacity(num);
        const date = try fmt.allocPrint(allocator, "{d} {s} {d} {d:0>2}:{d:0>2}:{d:0>2}", .{ t.year, t.month.name(), t.day, t.hour, t.minute, t.second });
        date_list.appendAssumeCapacity(date);

        max_row_len[0] = @max(max_row_len[0], num.len);
        max_row_len[1] = @max(max_row_len[1], (gen.description orelse "").len);
        max_row_len[2] = @max(max_row_len[2], date.len);
    }

    for (headers, 0..) |header, j| {
        var k: usize = 4 + max_row_len[j];

        const start_idx = (max_row_len[j] / 2) - (header.len / 2);

        var i: usize = 0;
        while (i < start_idx) : (i += 1) {
            log.print(" ", .{});
            k -= 1;
        }

        log.print("{s}", .{header});
        k -= header.len;

        while (k > 0) : (k -= 1) {
            log.print(" ", .{});
        }
    }
    log.print("\n", .{});
    for (max_row_len, 0..) |len, col| {
        var lim = len + 4;
        if (col == max_row_len.len - 1) {
            lim -= 4;
        }
        var i: usize = 0;
        while (i < lim) : (i += 1) {
            log.print("-", .{});
        }
    }
    log.print("\n", .{});

    for (generations, num_list.items, date_list.items, 0..) |gen, num, date, idx| {
        const row = [_][]const u8{ num, gen.description orelse "", date };
        for (row, 0..) |col, j| {
            log.print("{s}", .{col});
            var k: usize = max_row_len[j] - col.len;
            while (k > 0) {
                log.print(" ", .{});
                k -= 1;
            }
            if (j < row.len - 1) {
                log.print(" :: ", .{});
            }
        }
        if (idx < generations.len - 1) log.print("\n", .{});
    }
    log.print("\n\n", .{});
}

var exit_status: u8 = 0;

pub fn deleteGenerations(allocator: Allocator, generations: []GenerationMetadata, profile_dirname: []const u8) !void {
    var argv = ArrayList([]const u8).init(allocator);
    defer {
        for (argv.items) |s| allocator.free(s);
        argv.deinit();
    }

    try argv.append(try allocator.dupe(u8, "nix-env"));
    try argv.append(try allocator.dupe(u8, "-p"));
    try argv.append(try allocator.dupe(u8, profile_dirname));
    try argv.append(try allocator.dupe(u8, "--delete-generations"));

    for (generations) |gen| {
        const num = try fmt.allocPrint(allocator, "{d}", .{gen.generation.?});
        try argv.append(num);
    }

    log.cmd(argv.items);

    const result = runCmd(.{
        .allocator = allocator,
        .argv = argv.items,
        .stdout_type = .Ignore,
    }) catch return GenerationDeleteError.CommandFailed;

    if (result.status != 0) {
        exit_status = result.status;
        return GenerationDeleteError.CommandFailed;
    }
}

fn regenerateBootMenu(allocator: Allocator) !void {
    const argv = &.{ Constants.current_system ++ "/bin/switch-to-configuration", "boot" };

    const result = runCmd(.{
        .allocator = allocator,
        .argv = argv,
        .stdout_type = .Inherit,
    }) catch return GenerationDeleteError.CommandFailed;
    if (result.status != 0) {
        exit_status = result.status;
        return GenerationDeleteError.CommandFailed;
    }
}

pub fn generationDelete(allocator: Allocator, args: GenerationDeleteCommand, profile: ?[]const u8) GenerationDeleteError!void {
    if (linux.geteuid() != 0) {
        utils.execAsRoot(allocator) catch |err| {
            log.err("unable to re-exec this command as root: {s}", .{@errorName(err)});
            return GenerationDeleteError.PermissionDenied;
        };
    }

    const profile_name = profile orelse "system";
    const profile_dirname = if (mem.eql(u8, profile_name, "system"))
        "/nix/var/nix/profiles"
    else
        "/nix/var/nix/profiles/system-profiles";

    var generations_dir = fs.cwd().openDir(profile_dirname, .{ .iterate = true }) catch |err| {
        log.err("unable to open profile dir {s}: {s}", .{ profile_dirname, @errorName(err) });
        return GenerationDeleteError.ResourceAccessFailed;
    };
    defer generations_dir.close();

    const all_gens_info = utils.generation.gatherGenerationsFromProfile(allocator, profile_name) catch return GenerationDeleteError.ResourceAccessFailed;
    defer {
        for (all_gens_info) |*gen| gen.deinit();
        allocator.free(all_gens_info);
    }

    var all_gen_numbers = try ArrayList(usize).initCapacity(allocator, all_gens_info.len);
    defer all_gen_numbers.deinit();

    for (all_gens_info) |gen| {
        all_gen_numbers.appendAssumeCapacity(gen.generation.?);
    }

    if (all_gen_numbers.items.len < 2) {
        if (all_gen_numbers.items.len == 0) {
            log.err("no generations exist for profile {s}", .{profile_name});
        } else {
            log.err("only one generations exists for profile {s}; deletion is impossible", .{profile_name});
        }
        return;
    }

    // Make sure there are enough generations that exist.
    if (args.min) |min| {
        if (min >= all_gen_numbers.items.len) {
            log.err("there are {d} generations, but the expected minimum is {d}", .{ all_gen_numbers.items.len, min });
            log.info("keeping all generations", .{});
            return;
        }
    }

    var current_gen_number: usize = undefined;
    for (all_gens_info) |gen| {
        if (gen.current) {
            current_gen_number = gen.generation.?;
            if (mem.indexOf(usize, args.remove.items, &.{current_gen_number}) != null) {
                log.err("cannot remove generation {d}, this is the current generation!", .{current_gen_number});
                return GenerationDeleteError.InvalidParameter;
            }
        }
    }

    var gens_to_remove_set = std.AutoHashMap(usize, void).init(allocator);
    defer gens_to_remove_set.deinit();
    for (args.remove.items) |remove| {
        try gens_to_remove_set.put(remove, {});
    }

    var gens_to_keep_set = std.AutoHashMap(usize, void).init(allocator);
    defer gens_to_keep_set.deinit();
    for (args.keep.items) |keep| {
        try gens_to_keep_set.put(keep, {});
    }
    try gens_to_keep_set.put(current_gen_number, {});

    if (args.all) {
        for (all_gen_numbers.items) |gen| {
            try gens_to_remove_set.put(gen, {});
        }
    } else {
        // Add ranges
        if (args.from != null or args.to != null) {
            const upper_bound = args.to orelse all_gen_numbers.items[all_gen_numbers.items.len - 1];
            const lower_bound = args.from orelse all_gen_numbers.items[0];

            if (lower_bound > upper_bound) {
                log.err("lower bound '{d}' must be less than upper bound '{d}'", .{ lower_bound, upper_bound });
                return GenerationDeleteError.InvalidParameter;
            }

            // Make sure that ranges are within generation bounds
            if (upper_bound > all_gen_numbers.items[all_gen_numbers.items.len - 1] or upper_bound < all_gen_numbers.items[0]) {
                log.err("upper bound '{d}' is not within the range of available generations", .{upper_bound});
                return GenerationDeleteError.InvalidParameter;
            }
            if (lower_bound > all_gen_numbers.items[all_gen_numbers.items.len - 1] or lower_bound < all_gen_numbers.items[0]) {
                log.err("lower bound '{d}' is not within the range of available generations", .{lower_bound});
                return GenerationDeleteError.InvalidParameter;
            }

            for (all_gen_numbers.items) |gen| {
                if (gen >= lower_bound and gen <= upper_bound) {
                    try gens_to_remove_set.put(gen, {});
                }
            }
        }

        // Calulate generations before timestamp
        if (args.older_than) |older_than| {
            const now = zeit.instant(.{ .source = .now }) catch unreachable;
            const before_instant = zeit.instant(.{ .source = .{ .unix_nano = (now.timestamp - older_than.toEpochTime()) } }) catch unreachable;

            for (all_gen_numbers.items) |gen| {
                if (gens_to_remove_set.contains(gen)) {
                    continue;
                }
                const gen_name = try fmt.allocPrint(allocator, "system-{d}-link", .{gen});
                defer allocator.free(gen_name);

                const stat = generations_dir.statFile(gen_name) catch |err| {
                    log.err("unable to stat {s}: {s}; skipping date check", .{ gen_name, @errorName(err) });
                    continue;
                };

                if (stat.ctime < before_instant.timestamp) {
                    try gens_to_remove_set.put(gen, {});
                }
            }
        }
    }

    // Remove all members of gens_to_keep from gens_to_remove
    var keep_iter = gens_to_keep_set.keyIterator();
    while (keep_iter.next()) |keep| {
        _ = gens_to_remove_set.remove(keep.*);
    }

    // Ensure that the minimum number of generations exists.
    // Generations with higher numbers are prioritized.
    var remaining_number_of_generations = all_gen_numbers.items.len - gens_to_remove_set.count();

    if (args.min != null and remaining_number_of_generations < args.min.?) {
        const min_required = args.min.?;

        var j: usize = 0;
        while (j < all_gen_numbers.items.len - 1) : (j += 1) {
            const i = all_gen_numbers.items.len - 1 - j;
            const gen = all_gen_numbers.items[i];

            _ = gens_to_remove_set.remove(gen);

            remaining_number_of_generations = all_gen_numbers.items.len - gens_to_remove_set.count();
            if (remaining_number_of_generations == min_required) {
                break;
            }
        }
    }

    if (gens_to_remove_set.count() == 0) {
        log.warn("no generations were resolved for deletion from the given parameters", .{});
        log.info("there is nothing to do; exiting", .{});
        return;
    }

    var gens_to_remove_info = try ArrayList(GenerationMetadata).initCapacity(allocator, gens_to_remove_set.count());
    defer gens_to_remove_info.deinit();

    var remove_iter = gens_to_remove_set.keyIterator();
    while (remove_iter.next()) |gen_number| {
        for (all_gens_info) |gen| {
            if (gen.generation.? == gen_number.*) {
                gens_to_remove_info.appendAssumeCapacity(gen);
                break;
            }
        } else unreachable;
    }
    sort.block(GenerationMetadata, gens_to_remove_info.items, {}, GenerationMetadata.lessThan);

    try printDeleteSummary(allocator, gens_to_remove_info.items);
    log.print("There will be {d} generations left on this machine.\n", .{remaining_number_of_generations});

    if (!args.yes) {
        const confirm = utils.confirmationInput() catch |err| {
            log.err("unable to read stdin for confirmation: {s}", .{@errorName(err)});
            return GenerationDeleteError.ResourceAccessFailed;
        };
        if (!confirm) {
            log.warn("confirmation was not given, not proceeding", .{});
            return;
        }
    }

    log.step("Deleting generations...", .{});

    const full_profile_dirname = try fs.path.join(allocator, &.{ profile_dirname, profile_name });
    defer allocator.free(full_profile_dirname);

    try deleteGenerations(allocator, gens_to_remove_info.items, full_profile_dirname);

    log.step("Regenerating boot menu entries...", .{});

    try regenerateBootMenu(allocator);

    log.print("Success!\n", .{});
    log.info("to free up disk space from these generations, run `{s}`", .{if (opts.flake) "nix store gc" else "nix-collect-garbage"});
    if (!opts.flake) {
        log.info("if using `nix-collect-garbage`, the `-d` option frees up generations, which you may not want", .{});
    }
}

pub fn generationDeleteMain(allocator: Allocator, args: GenerationDeleteCommand, profile: ?[]const u8) u8 {
    generationDelete(allocator, args, profile) catch |err| {
        return switch (err) {
            GenerationDeleteError.ResourceAccessFailed => 3,
            GenerationDeleteError.InvalidParameter => 2,
            GenerationDeleteError.PermissionDenied => 13,
            GenerationDeleteError.CommandFailed => if (exit_status != 0) exit_status else 1,
            else => 1,
        };
    };

    return 0;
}

File: cmd/generation/delete/delete.go
package delete

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"github.com/olekukonko/tablewriter"
	"github.com/spf13/cobra"

	genUtils "github.com/water-sucks/nixos/cmd/generation/shared"
	buildOpts "github.com/water-sucks/nixos/internal/build"
	cmdTypes "github.com/water-sucks/nixos/internal/cmd/types"
	cmdUtils "github.com/water-sucks/nixos/internal/cmd/utils"
	"github.com/water-sucks/nixos/internal/constants"
	"github.com/water-sucks/nixos/internal/generation"
	"github.com/water-sucks/nixos/internal/logger"
	"github.com/water-sucks/nixos/internal/settings"
	"github.com/water-sucks/nixos/internal/system"
	timeUtils "github.com/water-sucks/nixos/internal/time"
	"github.com/water-sucks/nixos/internal/utils"
)

func GenerationDeleteCommand(genOpts *cmdTypes.GenerationOpts) *cobra.Command {
	opts := cmdTypes.GenerationDeleteOpts{}

	cmd := cobra.Command{
		Use:   "delete [flags] [GEN...]",
		Short: "Delete generations from this system",
		Long:  "Delete NixOS generations from this system.",
		Args: func(cmd *cobra.Command, args []string) error {
			for _, v := range args {
				value, err := strconv.ParseInt(v, 10, 32)
				if err != nil {
					return fmt.Errorf("[GEN] must be integer value, got '%v'", v)
				}
				opts.Remove = append(opts.Remove, uint(value))
			}
			if cmd.Flags().Changed("older-than") {
				// Make sure older-than is a valid systemd.time(7) string
				if _, err := timeUtils.DurationFromTimeSpan(opts.OlderThan); err != nil {
					return fmt.Errorf("invalid value for --older-than: %v", err.Error())
				}
			}

			for _, remove := range opts.Remove {
				for _, keep := range opts.Keep {
					if remove == keep {
						return fmt.Errorf("cannot remove and keep the same generation %v", remove)
					}
				}
			}

			log := logger.FromContext(cmd.Context())

			if opts.All {
				if opts.LowerBound != 0 {
					log.Warn("--all was specified, ignoring --from")
				}
				if opts.OlderThan != "" {
					log.Warn("--all was specified, ignoring --older-than")
				}
				if opts.UpperBound != 0 {
					log.Warn("--all was specified, ignoring --to")
				}
				if len(opts.Remove) != 0 {
					log.Warn("--all was specified, ignoring positional arguments")
				}
			}

			if !opts.All && opts.LowerBound == 0 && opts.UpperBound == 0 && len(opts.Remove) == 0 && opts.OlderThan == "" && len(opts.Keep) == 0 && opts.MinimumToKeep == 0 {
				return fmt.Errorf("no generations or deletion parameters were given")
			}

			return nil
		},
		ValidArgsFunction: generation.CompleteGenerationNumber(&genOpts.ProfileName, 0),
		RunE: func(cmd *cobra.Command, args []string) error {
			return cmdUtils.CommandErrorHandler(generationDeleteMain(cmd, genOpts, &opts))
		},
	}

	cmd.Flags().BoolVarP(&opts.All, "all", "a", false, "Delete all generations except the current one")
	cmd.Flags().Uint64VarP(&opts.LowerBound, "from", "f", 0, "Delete all generations after `gen`, inclusive")
	cmd.Flags().Uint64VarP(&opts.UpperBound, "to", "t", 0, "Delete all generations until `gen`, inclusive")
	cmd.Flags().Uint64VarP(&opts.MinimumToKeep, "min", "m", 0, "Keep a minimum of `num` generations")
	cmd.Flags().StringVarP(&opts.OlderThan, "older-than", "o", "", "Delete all generations older than `period`")
	cmd.Flags().UintSliceVarP(&opts.Keep, "keep", "k", nil, "Always keep this `gen`, can be specified many times")
	cmd.Flags().BoolVarP(&opts.Verbose, "verbose", "v", false, "Show verbose logging")
	cmd.Flags().BoolVarP(&opts.AlwaysConfirm, "yes", "y", false, "Automatically confirm generation deletion")

	err := cmd.RegisterFlagCompletionFunc("from", generation.CompleteGenerationNumberFlag(&genOpts.ProfileName))
	if err != nil {
		panic(err)
	}
	err = cmd.RegisterFlagCompletionFunc("to", generation.CompleteGenerationNumberFlag(&genOpts.ProfileName))
	if err != nil {
		panic(err)
	}
	err = cmd.RegisterFlagCompletionFunc("keep", generation.CompleteGenerationNumberFlag(&genOpts.ProfileName))
	if err != nil {
		panic(err)
	}

	cmdUtils.SetHelpFlagText(&cmd)
	cmd.SetHelpTemplate(cmd.HelpTemplate() + `
Arguments:
    [GEN]       Generation number

These options and arguments can be combined ad-hoc as constraints.

The 'period' parameter in --older-than is a systemd.time(7) span
(i.e. "30d 2h 1m"). Check the manual page for more information.
`)

	return &cmd
}

func generationDeleteMain(cmd *cobra.Command, genOpts *cmdTypes.GenerationOpts, opts *cmdTypes.GenerationDeleteOpts) error {
	log := logger.FromContext(cmd.Context())
	cfg := settings.FromContext(cmd.Context())
	s := system.NewLocalSystem(log)

	if !s.IsNixOS() {
		msg := "this command can only be run on NixOS systems"
		log.Error(msg)
		return fmt.Errorf("%v", msg)
	}

	if os.Geteuid() != 0 {
		err := utils.ExecAsRoot(cfg.RootCommand)
		if err != nil {
			log.Errorf("failed to re-exec command as root: %v", err)
			return err
		}
	}

	generations, err := genUtils.LoadGenerations(log, genOpts.ProfileName, false)
	if err != nil {
		return err
	}

	gensToDelete, err := resolveGenerationsToDelete(generations, opts)
	if err != nil {
		log.Errorf("%v", err)

		switch err.(type) {
		case GenerationResolveMinError:
			log.Info("keeping all generations")
		case GenerationResolveNoneFoundError:
			log.Info("there is nothing to do; exiting")
		}
		return err
	}

	remainingGenCount := len(generations) - len(gensToDelete)

	log.Print("The following generations will be deleted:")
	log.Print()
	displayDeleteSummary(gensToDelete)
	log.Printf("\nThere will be %v generations remaining on this machine.", remainingGenCount)
	log.Print()

	if !opts.AlwaysConfirm {
		confirm, err := cmdUtils.ConfirmationInput("Proceed?")
		if err != nil {
			log.Errorf("failed to get confirmation: %v", err)
			return err
		}
		if !confirm {
			log.Info("confirmation was not given, not proceeding")
			return nil
		}
	}

	log.Step("Deleting generations...")

	profileDirectory := generation.GetProfileDirectoryFromName(genOpts.ProfileName)
	if err := deleteGenerations(s, profileDirectory, gensToDelete, opts.Verbose); err != nil {
		log.Errorf("failed to delete generations: %v", err)
		return err
	}

	log.Step("Regenerating boot menu entries...")

	if err := regenerateBootMenu(s, opts.Verbose); err != nil {
		log.Errorf("failed to regenerate boot menu entries: %v", err)
		return err
	}

	log.Step("Collecting garbage...")

	if err := collectGarbage(s, opts.Verbose); err != nil {
		log.Errorf("failed to collect garbage: %v", err)
		return err
	}

	log.Print("Success!")

	return nil
}

func displayDeleteSummary(generations []generation.Generation) {
	data := make([][]string, len(generations))

	for i, v := range generations {
		data[i] = []string{
			fmt.Sprintf("%v", v.Number),
			v.Description,
			fmt.Sprintf("%v", v.CreationDate.Format(time.ANSIC)),
		}
	}

	table := tablewriter.NewWriter(os.Stdout)

	table.SetHeader([]string{"#", "Description", "Creation Date"})
	table.SetHeaderAlignment(tablewriter.ALIGN_CENTER)
	table.SetAlignment(tablewriter.ALIGN_LEFT)
	table.SetAutoFormatHeaders(false)
	table.SetAutoWrapText(false)
	table.SetBorder(false)
	table.SetRowSeparator("-")
	table.SetColumnSeparator("|")
	table.AppendBulk(data)
	table.Render()
}

func deleteGenerations(s system.CommandRunner, profileDirectory string, generations []generation.Generation, verbose bool) error {
	argv := []string{"nix-env", "-p", profileDirectory, "--delete-generations"}
	for _, v := range generations {
		argv = append(argv, fmt.Sprintf("%v", v.Number))
	}

	if verbose {
		s.Logger().CmdArray(argv)
	}

	cmd := system.NewCommand(argv[0], argv[1:]...)
	_, err := s.Run(cmd)
	return err
}

func regenerateBootMenu(s system.CommandRunner, verbose bool) error {
	switchToConfiguration := filepath.Join(constants.CurrentSystem, "bin", "switch-to-configuration")

	argv := []string{switchToConfiguration, "boot"}

	if verbose {
		s.Logger().CmdArray(argv)
	}

	cmd := system.NewCommand(argv[0], argv[1:]...)
	_, err := s.Run(cmd)
	return err
}

func collectGarbage(s system.CommandRunner, verbose bool) error {
	var argv []string
	if buildOpts.Flake == "true" {
		argv = []string{"nix", "store", "gc"}
	} else {
		argv = []string{"nix-collect-garbage"}
	}

	if verbose {
		argv = append(argv, "-v")
		s.Logger().CmdArray(argv)
	}

	var cmd *system.Command
	if len(argv) == 1 {
		cmd = system.NewCommand(argv[0])
	} else {
		cmd = system.NewCommand(argv[0], argv[1:]...)
	}

	_, err := s.Run(cmd)
	return err
}
